README.md
# Traktor V2.5
Traktor will autamically install Tor, privoxy, dnscrypt-proxy, resolvconf and Tor Browser Launcher in either a Debian based distro like Ubuntu, an Arch based distro, Fedora based distro or an OpenSUSE based distro and configures them as well.

To do this, just run 'installer.sh' file in a supported shell like bash and watch for prompts it asks you.

## Note
**Currently, Fedora and Arch aren't supported by Traktor.**

Do NOT expect anonymity using this method. Privoxy is an http proxy and can leak data. If you need anonymity or strong privacy, manually run torbrowser-launcher after installing traktor and use it.

### TODO
- [x] Dekstop Installer
    - [x] Indicator
    - [x] Support TorDNS
    - [x] Support OpenDNS
    - [x] Gnome Extension
    - [ ] Personal Package Archives (PPA)
- [x] Server Installer

### Support Distro
- [x] Desktop
    - [x] Ubuntu (16.04 | 17.10 | 18.04)
    - [x] Debian
    - [x] OpenSUSE
    - [x] Arch (not any more)
    - [x] Fedora (not any more)
- [x] Server
    - [x] Ubuntu
        - [x] Ubuntu Server Raspberry Pi (18.04)

### Support Dekstop
- [x] Gnome
- [x] Unity 7
- [x] XFCE
- [x] KDE
- [x] LXDE
- [x] Mate

### Support Bridge
- [x] obfs3
    - [ ] for Orbot
- [x] obfs4
    - [x] for [Orbot](https://gitlab.com/GNULand/TraktorPlus/Traktor/raw/config/orbot-obfs4-bridges)
- [ ] meek
    - [ ] for Orbot

## Manual Pages
[Traktor Manual Page](https://gitlab.com/GNULand/TraktorPlus/Traktor/wikis/Traktor-manual-page-(English)) {lang=EN}

[راهنمای اسکریپت تراکتور به زبان فارسی](https://gitlab.com/GNULand/TraktorPlus/Traktor/wikis/Traktor-manual-page-(persian)) {lang=FA}

## Install
```bash
$ sudo apt install net-tools
$ wget https://github.com/TraktorPlus/Traktor/archive/master.zip -O Traktor.zip
$ unzip Traktor.zip -d $HOME/Traktor && cd $HOME/Traktor/*
$ chmod +x installer.sh
$ ./installer.sh
```

### torrc Generator
New option for share Tor Proxy on your local Network. {see [Manual Pages](#manual-pages) for more Information}
```bash
$ chmod +x torrc-gen.py
$ ./torrc-gen.py
```
If you Like use this option on your Mobile or PC, type your Local IP Network instead `LocalIpMachine` in "Proxy Settings or Network Settings".

HTTP:
`LocalIpMachine:8181`

SOCKS:
`LocalIpMachine:9050`

> Example:

> My Local IP Network: 192.168.1.2

> for HTTP: `192.168.1.2:8181`

> for SOCKS: `192.168.1.2:9050`


### Other for Traktor Panel (Indicator)
[Traktor Panel:](https://gitlab.com/GNULand/TraktorPlus/Traktor-Panel)

![panel](Traktor-Panel.png)

1. Install the dependencies with your package manager. For instance, on Ubuntu:
```bash
$ sudo apt install curl
$ sudo apt install python-gi python-appindicator gir1.2-appindicator3-0.1
```

2. Clone the script repository and Run the script:
```bash
$ wget https://gitlab.com/GNULand/TraktorPlus/Traktor-Panel/repository/master/archive.zip -O Traktor-Panel.zip
$ unzip Traktor-Panel.zip -d $HOME/Traktor-Panel && cd $HOME/Traktor-Panel/*
$ chmod +x installer-panel.sh
$ ./installer-panel.sh
```

### Other for Traktor Extension (Gnome Extension)
[Traktor Extension:](https://gitlab.com/GNULand/TraktorPlus/Traktor-Gnome-Extension)

![extension](Traktor-Extension.png)
```bash
$ wget https://gitlab.com/GNULand/TraktorPlus/Traktor-Gnome-Extension/repository/master/archive.zip -O Traktor-Gnome-Extension.zip
$ unzip Traktor-Gnome-Extension.zip -d $HOME/Traktor-Gnome-Extension && cd $HOME/Traktor-Gnome-Extension/*
$ make build
$ sudo make install
```

## Changes
[See Changes](https://gitlab.com/GNULand/TraktorPlus/Traktor/blob/master/CHANGELOG)

## Troubleshooting
[See Troubleshooting](https://gitlab.com/GNULand/TraktorPlus/Traktor/wikis/Troubleshooting)

## Our Team
- [Sosha](https://gitlab.com/Sosha)
- [shervinmo](https://gitlab.com/shervinmoheban)
- [Pouya Abbasian](https://gitlab.com/pouya-abbasian)

### Special Thanks to
- [Mostafa Asadi ](https://gitlab.com/mostafaasadi)
- [Moein](https://gitlab.com/moeinroid)
- [Ehsaan](https://gitlab.com/Ehsaan)
- [alipg](https://gitlab.com/alipg)


CHANGELOG
2017-04-30 Version 1.6
    1. fix tor icons
    2. add new obfs3 bridges
    3. add new obfs4 bridges
    4. backup torrc file
    5. added privoxy package
    6. remove polipo package
    7. change http port
    
2017-05-30 Verion 1.7
    1. traktor is now supporting fedora
    2. add traktor_fedora.sh file
    3. remove unistall_debian.sh file
    4. remove traktor_gui_panel folder
    
2017-07-10 Version 1.8
    1. traktor is now supporting OpenSUSE
    2. add traktor_opensuse.sh
    
2017-08-06 Version 1.9
    1. add uninstall_debian.sh file for Removed Packages and Configured
    2. v uninstall_opensuse.sh file for Removed Packages and Configured
    3. edit show message
    4. fix NetworkManager.conf
    5. bug fixes
    6. fix important issues
    
2017-11-08 Version 2
    1. add fantasy installing process for traktor_debian.sh
    2. add new obfs3 bridges
    3. add new obfs4 bridges
    4. add .Traktor folder on your HOME
    5. add Traktor Panel (System Tray Icon)
    6. add installer.sh file
    7. add traktor file
    8. traktor is now supporting Gnome, XFCE and KDE Desktop
    9. add '--update' Option Command for Update torrc
    10. remove traktor.sh file 
    11. control traktor with command-line by Swtich
    12. edited torrc file
    13. Change README.md
    14. Change CHANGELOG
    15. bug fixes
    16. fix important issues
    17. add more bugs to fix later
    
2017-12-27 Version 2.1
    1. fix apt problem when found new Update for tor and apparmor packages
    2. fix bug installer.sh file
    3. add fantasy installing process for traktor_opensuse.sh
    4. smarter Installation
    5. traktor_opensuse is now supporting KDE
    6. add dns-fixer.sh file

2018-01-16 Version 2.2
    1. traktor is now supporting Unity and LXDE Desktop
    2. add resolvconf package
    3. add TorDNS and OpenDNS to resolv.conf
    4. add new uninstall version for traktor_debian.sh
    5. add new obfs3 bridges
    6. add new obfs4 bridges
    7. add dns.log file
    8. remove dns-fixer.sh file
    9. Change README.md
    10. Change CHANGELOG
    11. fix apt bug
    12. fix important issues
    13. bug fixes
    14. edited torrc file
    15. Currently, Fedora isn't supported by Traktor
    16. Currently, Arch isn't supported by Traktor
    17. added more bugs to fix later
    
2018-02-14 Version 2.3
    1. new Traktor Panel (indicator [or system try icon])
    2. change Traktor Panel icons
    3. fix torrc bug
    4. fix libssl address
    5. added more bugs to fix later
    
2018-03-05 Version 2.4
    1. Trakor is now have a Gnome Extension
    2. New uninstall_debian.sh
    3. fix traktor_debian.sh bug
    4. add "rslv" function in traktor_debian.sh
    5. remove "panel" function in traktor_debian.sh
    6. remove Traktor-Panel folder
    7. add more bugs to fix later
    
2018-04-19 Version 2.5
    1. fix unity support
    2. fix torrc bug
    3. fix boot function
    4. change Arch Installer
    5. change Fedora Installer
    6. change README.md
    7. change installer.sh
    8. chanfe trktor_debian.sh
    9. add traktor-plus.sh
    10. add torrc-gen.py
    11. add traktor_raspberry.sh
    12. add Wiki on GitLab


traktor_debian.sh
#!/bin/bash


##################################################
#                  Traktor V2.5                  #
#     https://github.com/TraktorPlus/Traktor     #
# https://gitlab.com/GNULand/TraktorPlus/Traktor #
##################################################


trap "exit 1" TERM
export TOP_PID=$$

clear
Version="Traktor V2.5"
#date="2018/06/12 Tue 08:00"
echo -e "$Version\n    Tor will be automatically installed and configured…\n\n"


# =========/Create File and Folder/=========

if [ ! -e "$HOME/.Traktor/Traktor_Log/install.log" ]; then
    echo -e " $Version Log: [$(date)]" | tee $HOME/.Traktor/Traktor_Log/install.log > /dev/null # echo date and time
fi

# =========/Create File and Folder/=========


# =========/Color/=========

GRE='\033[92m' # Green Light
RD='\033[91m' # Red Light
YLW='\033[93m' # Yellow Light
NC='\033[0m' # White

# =========\Color\=========


# =========\Functions\=========

function loading {
    spin[0]="-"
    spin[1]="\\"
    spin[2]="|"
    spin[3]="/"
    PID=$!
    echo -n "${spin[0]}"
    while [ -d /proc/$PID ]
    do
        for i in "${spin[@]}"
            do
            echo -ne "\b$i"
            sleep 0.2
        done
    done
    wait $PID
    if [ $? -ne 0 ]; then
        echo -e "\b${RD}Failed!"
        echo -e "\n${NC}Check Traktor Log in $HOME/.Traktor/Traktor_Log/install.log"
        kill -s TERM $TOP_PID
    else
        echo -e "\b${GRE}Done."
    fi
}

function updte {
    if ! { sudo apt-get update 2>&1 >> $HOME/.Traktor/Traktor_Log/install.log || echo E: update failed; } | grep -q '^[WE]:'; then
        echo success
    else
        echo failure
        exit 1
    fi
    echo ""
}

function packinOne { 
    sudo apt-get --yes --force-yes -o Dpkg::Options::="--force-confnew" install \
    tor \
    obfs4proxy
	if [ "$?" != "0" ]; then
        exit 1
	fi
}

function rslv {
    chckrslv=`apt list --installed | grep '^resolvconf/'`
    if [[ "$?" != "0" ]]; then
        echo -e "#resolvconf Not Installed!"
    fi
}

function packinTwo {
    rslv
    
    sudo apt-get install -y \
    dnscrypt-proxy \
    resolvconf \
    privoxy \
    torbrowser-launcher \
    apt-transport-tor
    if [ "$?" != "0" ]; then
        exit 1
    fi
}

function torrcbackup {
    if [ -e "/etc/tor/torrc" ]; then
        sudo cp /etc/tor/torrc /etc/tor/torrc.traktor-backup
        if [ "$?" != "0" ]; then
            exit 1
        fi
    fi

}

function bridgewrite {
    sudo wget https://raw.githubusercontent.com/TraktorPlus/Traktor/config/torrc -O /etc/tor/torrc
    if [ "$?" != "0" ]; then
        exit 1
    fi
}

function fixappar { 
        sudo sed -i '27s/PUx/ix/' /etc/apparmor.d/abstractions/tor
        if [ "$?" != "0" ]; then
            exit 1 
        fi
        sudo apparmor_parser -r -v /etc/apparmor.d/system_tor
        if [ "$?" != "0" ]; then
            exit 1 
        fi
}

function privoxyback {
    if [ -e "/etc/privoxy/config" ]; then
        sudo cp /etc/privoxy/config /etc/privoxy/config.traktor-backup
        if [ "$?" != "0" ]; then
            exit 1
        fi
    fi
}

function privoxyconf {
    sudo perl -i -pe 's/^listen-address/#$&/' /etc/privoxy/config
    cmd1="$?"
    echo 'logdir /var/log/privoxy
listen-address  0.0.0.0:8118
forward-socks5t             /     127.0.0.1:9050 .
forward         192.168.*.*/     .
forward            10.*.*.*/     .
forward           127.*.*.*/     .
forward           localhost/     .' | sudo tee /etc/privoxy/config
    cmd2="$?"
    echo ""
    sudo systemctl enable privoxy
    cmd3="$?"
    sudo systemctl restart privoxy.service
    cmd4="$?"
    echo ""
    if [ "$cmd1" != "0" ] || [ "$cmd2" != "0" ] || [ "$cmd3" != "0" ] || [ "$cmd4" != "0" ]; then
        exit 1
    fi
}

function stopnet {
    sudo sed -i -e 's/\[main\]/\[main\]\ndns=none/g' /etc/NetworkManager/NetworkManager.conf
    if [ "$?" != "0" ]; then
        exit 1
    fi
}

function chkowner {
    sudo chown debian-tor:adm /var/log/tor/notices.log
    if [ "$?" != "0" ]; then
        exit 1
    else
        echo "change Ownership notices.log"
    fi
}

function dns {
    org_ip="127.0.0."
    for number in {10..255} ; do
    ip=$org_ip$number
    
    logfile="$HOME/.Traktor/Traktor_Log/dns.log"

    cat $logfile | grep $ip > /dev/null

    if [ "$?" == "0" ]; then # Check Log if Positive
        cat /etc/resolv.conf | grep $ip > /dev/null
        if [ "$?" == "0" ]; then # Check resolve.conf if Positive
            if { sudo netstat -antpou 2>&1; } | grep "$ip:53" | awk {'print $6'} | grep tor > /dev/null; then # Check netstat for TorDNS
                sudo sed -i '/# TorDNS/d' /etc/tor/torrc
                sudo sed -i '/DNSPort/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsOnResolve/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsSuffixes/d' /etc/tor/torrc
                echo -e "\n# TorDNS" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "DNSPort $ip:53" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsOnResolve 1" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsSuffixes .exit,.onion" | sudo tee -a /etc/tor/torrc > /dev/null
                echo -e "TorDNS Running Already( Log OK, RC OK, Net OK )\n"
                break
            elif { sudo netstat -antpou 2>&1; } | grep "$ip:53"; then
                continue
            else
                sudo sed -i '/# TorDNS/d' /etc/tor/torrc
                sudo sed -i '/DNSPort/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsOnResolve/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsSuffixes/d' /etc/tor/torrc
                echo -e "\n# TorDNS" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "DNSPort $ip:53" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsOnResolve 1" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsSuffixes .exit,.onion" | sudo tee -a /etc/tor/torrc > /dev/null
                echo -e "TorDNS Activated( Log OK, RC OK, Net NO! )\n"
                break
            fi
        else # Check resolve.conf if Negative
            if { sudo netstat -antpou 2>&1; } | grep "$ip:53" | awk {'print $6'} | grep tor > /dev/null; then # Check netstat for TorDNS
                echo "nameserver $ip" | sudo tee -a /etc/resolvconf/resolv.conf.d/head
                echo "nameserver 208.67.222.222" | sudo tee -a /etc/resolvconf/resolv.conf.d/tail # added OpenDNS
                sudo resolvconf -u
                echo -e "TorDNS Activated( Log OK, RC NO!, Net OK )\n"
                break
            elif { sudo netstat -antpou 2>&1; } | grep "$ip:53"; then
                continue
            else
                echo "nameserver $ip" | sudo tee -a /etc/resolvconf/resolv.conf.d/head
                echo "nameserver 208.67.222.222" | sudo tee -a /etc/resolvconf/resolv.conf.d/tail # added OpenDNS
                sudo resolvconf -u
                sudo sed -i '/# TorDNS/d' /etc/tor/torrc
                sudo sed -i '/DNSPort/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsOnResolve/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsSuffixes/d' /etc/tor/torrc
                echo -e "\n# TorDNS" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "DNSPort $ip:53" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsOnResolve 1" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsSuffixes .exit,.onion" | sudo tee -a /etc/tor/torrc > /dev/null
                echo -e "TorDNS Activated( Log OK, RC NO!, Net NO! )\n"
                break
            fi
        fi
    else #Check Log if Negative
        cat /etc/resolv.conf | grep $ip > /dev/null
        if [ "$?" == "0" ]; then
            if { sudo netstat -antpou 2>&1; } | grep "$ip:53" | awk {'print $6'} | grep tor > /dev/null; then # Check netstat for TorDNS
                echo -e "\nDNS Log: [$(date)]:\n$ip" | tee -a $HOME/.Traktor/Traktor_Log/dns.log > /dev/null
                echo -e "TorDNS Running Already( Log NO!, RC OK, Net OK )\n"
                break
            elif { sudo netstat -antpou 2>&1; } | grep "$ip:53"; then
                continue
            else
                sudo sed -i '/# TorDNS/d' /etc/tor/torrc
                sudo sed -i '/DNSPort/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsOnResolve/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsSuffixes/d' /etc/tor/torrc
                echo -e "\n# TorDNS" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "DNSPort $ip:53" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsOnResolve 1" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsSuffixes .exit,.onion" | sudo tee -a /etc/tor/torrc > /dev/null
                echo -e "\nDNS Log: [$(date)]:\n$ip" | tee -a $HOME/.Traktor/Traktor_Log/dns.log > /dev/null
                echo -e "TorDNS Activated( Log NO!, RC OK, Net NO! )\n"
                break
            fi
        else
            if { sudo netstat -antpou 2>&1; } | grep "$ip:53" | awk {'print $6'} | grep tor > /dev/null; then # Check netstat for TorDNS
                echo "nameserver $ip" | sudo tee -a /etc/resolvconf/resolv.conf.d/head
                echo "nameserver 208.67.222.222" | sudo tee -a /etc/resolvconf/resolv.conf.d/tail # added OpenDNS
                sudo resolvconf -u
                echo -e "\nDNS Log: [$(date)]:\n$ip" | tee -a $HOME/.Traktor/Traktor_Log/dns.log > /dev/null
                echo -e "TorDNS Activated( Log NO!, RC NO!, Net OK )\n"
                break
            elif { sudo netstat -antpou 2>&1; } | grep "$ip:53"; then
                continue
            else
                echo "nameserver $ip" | sudo tee -a /etc/resolvconf/resolv.conf.d/head
                echo "nameserver 208.67.222.222" | sudo tee -a /etc/resolvconf/resolv.conf.d/tail # added OpenDNS
                sudo resolvconf -u
                sudo sed -i '/# TorDNS/d' /etc/tor/torrc
                sudo sed -i '/DNSPort/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsOnResolve/d' /etc/tor/torrc
                sudo sed -i '/AutomapHostsSuffixes/d' /etc/tor/torrc
                echo -e "\n# TorDNS" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "DNSPort $ip:53" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsOnResolve 1" | sudo tee -a /etc/tor/torrc > /dev/null
                echo "AutomapHostsSuffixes .exit,.onion" | sudo tee -a /etc/tor/torrc > /dev/null
                echo -e "\nDNS Log: [$(date)]:\n$ip" | tee -a $HOME/.Traktor/Traktor_Log/dns.log > /dev/null
                echo -e "TorDNS Activated( Log NO!, RC NO!, Net NO! )\n"
                break
            fi
        fi
    fi

    number=$((number+1))
    done
}

function boot {
    sudo service tor restart
    if [ $? != "0" ]; then
    	exit 1
    fi
    sudo echo " " | sudo tee /var/log/tor/notices.log
    if [ $? != "0" ]; then
    	exit 1
    fi
    timer=1
    while [[ -e ./no && $timer -lt 120  ]]; do
       	if sudo cat /var/log/tor/notices.log | grep "Bootstrapped 100%: Done" > /dev/null; then
		rm ./no
		echo -e "\nBootstrapped 100%" >> $HOME/.Traktor/Traktor_Log/install.log
	else
		sleep 1
		timer=$((timer+1))
 	fi
    done
}

function libevent {
        read -p "libevent-2.0-5 not installed, do you want to install it? [Y/n] : " qa
        if [ "$qa" == "y" ] || [ "$qa" == "Y" ] || [ "$qa" == "" ]; then
            wget "http://ftp.de.debian.org/debian/pool/main/libe/libevent/libevent-2.0-5_2.0.21-stable-3_amd64.deb" -P $HOME/.Traktor/
            cmd1=$?
            sudo dpkg -i $HOME/.Traktor/libevent-2.0-5_2.0.21-stable-3_amd64.deb
            cmd2=$?
            if [ "$cmd1" != "0" ] || [ "$cmd2" != "0" ]; then
                exit 1
            fi
        else
            echo "Abort."
        fi
}

function libssl {
        read -p "libssl1.1 not installed, do you want to install it? [Y/n]: " qa
        if [ "$qa" == "y" ] || [ "$qa" == "Y" ] || [ "$qa" == "" ]; then
            wget "http://security.debian.org/debian-security/pool/updates/main/o/openssl/libssl1.1_1.1.0f-3+deb9u2_amd64.deb" -P $HOME/.Traktor/
            cmd1=$?
            sudo dpkg -i $HOME/.Traktor/libssl1.1_1.1.0f-3+deb9u1_amd64.deb
            cmd2=$?
            if [ "$cmd1" != "0" ] || [ "$cmd2" != "0" ]; then
                exit 1
            fi
        else
            echo "Abort."
        fi
}

function addrepo {
    codename=`lsb_release -c | awk {'print $2'}`
    if [[ "$codename" =~ ^(trusty|xenial|zesty|artful|bionic|wheezy|jessie|stretch|buster|sid|experimental)$ ]]; then
        echo "deb tor+http://deb.torproject.org/torproject.org $codename main" | sudo tee /etc/apt/sources.list.d/tor.list
        cmd1=$?
        echo "deb-src tor+http://deb.torproject.org/torproject.org $codename main" | sudo tee -a /etc/apt/sources.list.d/tor.list
        cmd2=$?
        echo "deb tor+http://deb.torproject.org/torproject.org obfs4proxy main" | sudo tee -a /etc/apt/sources.list.d/tor.list
        cmd3=$?
        echo ""
    else
        echo "deb tor+http://deb.torproject.org/torproject.org stable main" | sudo tee /etc/apt/sources.list.d/tor.list
        cmd4=$?
        echo "deb tor+http://deb.torproject.org/torproject.org obfs4proxy main" | sudo tee -a /etc/apt/sources.list.d/tor.list
        cmd5=$?
        echo ""
        # Check Package
        libsl=`apt list --installed 2> /dev/null | grep libssl1.1`
        if [[ "$?" == "0" ]]; then
            libsl_status=true
        else
            libsl_status=false
        fi

        libevnt=`apt list --installed 2> /dev/null | grep libevent-2.0-5`
        if [[ "$?" == "0" ]]; then
            libevnt_status=true
        else
            libevnt_status=false
        fi
        
        if [[ "$libsl_status" == "false" ]] || [[ "$libevnt_status" == "false" ]]; then
            echo -ne "\n${NC}Adding libssl or libevent...   "
            apt list --installed 2> /dev/null | grep libssl >> $HOME/.Traktor/Traktor_Log/install.log
            echo "" | tee -a $HOME/.Traktor/Traktor_Log/install.log
            apt list --installed 2> /dev/null | grep libevent >> $HOME/.Traktor/Traktor_Log/install.log
            if [[ "$libsl_status" == "false" ]]; then
                libssl
            fi
            if [[ "$libevnt_status" == "false" ]]; then
                libevent
            fi
            echo ""
        fi
    fi
}

function gpgkey {
    gpg --keyserver keys.gnupg.net --recv 886DDD89
    cmd1=$?
    gpg --export A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89 | sudo apt-key add -
    cmd2=$?
    echo ""
    if [ "$cmd1" != "0" ] || [ "$cmd2" != "0" ]; then
        exit 1
    fi
}

function updatemain {
    if ! { sudo apt-get update 2>&1 >> $HOME/.Traktor/Traktor_Log/install.log || echo E: update failed; } | grep -q '^[WE]:'; then
        echo success
    else
        echo failure
        exit 1
    fi
    echo ""
}

function upgrademain {
    sudo sudo apt-get --yes --force-yes -o Dpkg::Options::="--force-confnew" install \
    tor \
    obfs4proxy
    checktorrc=`sudo ls /etc/tor/torrc.dpkg-old &> /dev/null`
    if [ "$?" == "0" ]; then
        sudo cat /etc/tor/torrc.dpkg-old | sudo tee /etc/tor/torrc
    fi
    echo ""
}

function setip {
    echo $XDG_CURRENT_DESKTOP | grep -i gnome > /dev/null
    gnome=$?
    echo $XDG_CURRENT_DESKTOP | grep -i unity > /dev/null
    unity=$?
    echo $XDG_CURRENT_DESKTOP | grep -i ubuntu > /dev/null
    ubuntu=$?
    echo $XDG_CURRENT_DESKTOP | grep -i xfce > /dev/null
    xfce="$?"
    echo $XDG_CURRENT_DESKTOP | grep -i kde > /dev/null
    kde="$?"
    echo $XDG_CURRENT_DESKTOP | grep -i lxde > /dev/null
    lxde="$?"
    if [ "$gnome" == "0" ] || [ "$unity" == "0" ] || [ "$xfce" == "0" ] || [ "$ubuntu" == "0" ] || [ "$lxde" == "0" ]; then 
        gsettings set org.gnome.system.proxy mode 'manual'
        gsettings set org.gnome.system.proxy.http host 127.0.0.1
        gsettings set org.gnome.system.proxy.http port 8118
        gsettings set org.gnome.system.proxy.socks host 127.0.0.1
        gsettings set org.gnome.system.proxy.socks port 9050
        gsettings set org.gnome.system.proxy ignore-hosts "['localhost', '127.0.0.0/8', '::1', '192.168.0.0/16', '192.168.8.1', '10.0.0.0/8', '172.16.0.0/12', '0.0.0.0/8', '10.0.0.0/8', '100.64.0.0/10', '127.0.0.0/8', '169.254.0.0/16', '172.16.0.0/12', '192.0.0.0/24', '192.0.2.0/24', '192.168.0.0/16', '192.88.99.0/24', '198.18.0.0/15', '198.51.100.0/24', '203.0.113.0/24', '224.0.0.0/3']"
    elif [ "$kde" == "0" ]; then
        sed -i -- 's/ProxyType=.*/ProxyType=1/g' $HOME/.config/kioslaverc
        sed -i -- 's/httpProxy=.*/httpProxy=http:\/\/127.0.0.1 8118/g' $HOME/.config/kioslaverc
        sed -i -- 's/socksProxy=.*/socksProxy=socks:\/\/127.0.0.1 9050/g' $HOME/.config/kioslavercs
    else
        echo "Your Desktop not Support!"
    fi
}

# =========/End Functions/=========


# =========\Call Functions\=========

sudo uname -a &> /dev/null

echo -e "\n\n\n[ #install ] $Version Install Log: [$(date)]\n" | tee -a $HOME/.Traktor/Traktor_Log/install.log > /dev/null # echo date and time

echo -ne "${NC}Updating Repository...  "
updte >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Installing packages tor and obfs4proxy...  "
packinOne >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Installing Other packages...  "
packinTwo >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Backing up the old torrc to '/etc/tor/torrc.traktor-backup'...   "
torrcbackup >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Writing bridges...   "
bridgewrite >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

linecheck=`cat /etc/apparmor.d/abstractions/tor | head -19 | tail -1 | awk {'print $2$3$4'}`
if [ "$linecheck" != "Neededbyobfs4proxy" ]; then
    echo -ne "${NC}Fixing apparmor...   "
    fixappar >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
    loading
fi

echo -ne "${NC}Backing up privoxy...   "
privoxyback >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Config for privoxy...   "
privoxyconf >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}stop Network Manager from adding dns-servers to /etc/resolv.conf...   "
stopnet >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Set DNS Tor and OpenDNS...   "
dns >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

# Fix Owner
logfile=`sudo ls /var/log/tor/notices.log &> /dev/null`
if [ "$?" != "0" ]; then
    sudo touch /var/log/tor/notices.log
    echo -e "Create notices.log" >> $HOME/.Traktor/Traktor_Log/install.log
fi
chkdebtor=`sudo ls -lha /var/log/tor/notices.log | awk {'print $3'}`
chkadm=`sudo ls -lha /var/log/tor/notices.log | awk {'print $4'}`
if [[ "$chkdebtor" != "debian-tor" ]] || [[ "$chkadm" != "adm" ]]; then
    echo -ne "${NC}Fix Owner /var/log/tor/notices.log...   "
    chkowner >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
    loading
fi

counter=1
touch ./no
while [[ -e ./no && $counter -lt 3 ]]; do
	echo -ne "${NC}Tor is trying to establish a connection. This may take long for some minutes. Please wait...   ${YLW}"$counter"/2 Try... "
	boot
	counter=$((counter+1))
	loading
done
if [ -e ./no ]; then
	echo -e "\b${RD}Failed!"
    echo -e "\n${NC}Check Traktor Log in $HOME/.Traktor/Traktor_Log/install.log"
    kill -s TERM $TOP_PID
fi

echo -ne "${NC}Adding tor repos...   "
addrepo >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Fetching Tor signing key and adding it to the keyring...   "
gpgkey >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Updating tor from main repo...   "
updatemain >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Upgrading tor and obfs4proxy pakages...   "
upgrademain >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -ne "${NC}Set ip and port...   "
setip >> $HOME/.Traktor/Traktor_Log/install.log 2>&1 &
loading

echo -e "Install Complete!" >> $HOME/.Traktor/Traktor_Log/install.log

sleep 3

# =========/End Call Functions/=========

echo -e "${GRE}\n\nCongratulations!!! Your computer is using Tor.\n Now Use this command 'traktor --help' :)\n\n${NC}"

# Excute Traktor Plus
chmod +x traktor-plus.sh
./traktor-plus.sh
